node: device-mgr: Handle recovery flow by checking if healthy devices exist

swatisehgal opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug


#### What this PR does / why we need it:

In case of node reboot/kubelet restart, the flow of events involves obtaining the state from the checkpoint file followed by setting the `healthDevices`/`unhealthyDevices` to its zero value. This is done to allow the device plugin to re-register itself so that capacity can be updated appropriately.

During the allocation phase, we need to check if the resources requested by the pod have been registered AND healthy devices are present on the node to be allocated.

Also we need to move this check above `needed==0` where needed is required - devices allocated to the container (which is obtained from the checkpoint file) because even in cases where no additional devices have to be allocated (as they were pre-allocated), we still need to make sure that the devices that were previously allocated are healthy.

For more details refer to the comment here: https://github.com/kubernetes/kubernetes/issues/109595#issuecomment-1359882423.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #109595 and (possibly) #107928 (TODO: need to check the latter)
#### Special notes for your reviewer:

**NOTE: This PR depends on sample device plugin changes proposed in: https://github.com/kubernetes/kubernetes/pull/115107**. 
Those changes are included here (first two commits) as they are needed for e2e test implementation.

NOTE: On node reboot, in case the application pod requesting devices starts after the device plugin pod providing those devices as has been described in https://github.com/kubernetes/kubernetes/issues/109595 and https://github.com/kubernetes/kubernetes/issues/107928 we exit early as there are pre-allocated devices without checking if the devices have registered and are healthy. This PR adds both those checks.

Because of this change, in cases where application pods are started before device plugin pod (say after node reboot),  because devices are not healthy, the pod would fail with `UnexpectedAdmissionError` error.  If the pod is part of a deployment, another pod would be created but that would stay in `Pending` state waiting for the pod to be scheduled to the node. This pod would go `Running` after the node capacity is updated followed by start up of device plugin pod and its registration.

The pod which fails at admission time continues to exist on the node and needs to be removed manually:
```
kubectl delete pods --field-selector status.phase=Failed
```

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

original release note:
```
During device plugin allocation, resources requested by the pod can only be allocated if the device plugin has registered itself to kubelet AND healthy devices are present on the node to be allocated. If these conditions are not sattsfied, the pod would fail with `UnexpectedAdmissionError` error.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

swatisehgal committed:
node: device-mgr: Handle recovery by checking if healthy devices exist

In case of node reboot/kubelet restart, the flow of events involves
obtaining the state from the checkpoint file followed by setting
the `healthDevices`/`unhealthyDevices` to its zero value. This is
done to allow the device plugin to re-register itself so that
capacity can be updated appropriately.

During the allocation phase, we need to check if the resources requested
by the pod have been registered AND healthy devices are present on
the node to be allocated.

Also we need to move this check above `needed==0` where needed is
required - devices allocated to the container (which is obtained from
the checkpoint file) because even in cases where no additional devices
have to be allocated (as they were pre-allocated), we still need to
make the devices that were previously allocated are healthy.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

fromanirh commented:
/triage accepted
/priority important-longterm

fromanirh commented:
partially replaces #110729 - I'm going to close the latter when tests are added here

swatisehgal committed:
node: device-mgr: unit-tests: admission failure due to unhealthy devices

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal commented:
/retest

swatisehgal committed:
node: device-mgr: e2e: Implement End to end test

This commit reuses e2e tests implmented as part of https://github.com/kubernetes/kubernetes/pull/110729.
The commit is borrowed from the aforementioned PR as is to preserve
authorship. Subsequent commit will update the end to end test to
simulate the problem this PR is trying to solve by reproducing
the issue: 109595.

Co-authored-by: Francesco Romani <fromani@redhat.com>
Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: device-mgr: e2e: Update the e2e test to reproduce issue:109595

Breakdown of the steps implemented as part of this e2e test is as follows:
1. Create a file `registration` at path `/var/lib/kubelet/device-plugins/sample/`
2. Create sample device plugin with an environment variable with
   `REGISTER_CONTROL_FILE=/var/lib/kubelet/device-plugins/sample/registration` that
    waits for a client to delete the control file.
3. Trigger plugin registeration by deleting the abovementioned directory.
4. Create a test pod requesting devices exposed by the device plugin.
5. Stop kubelet.
6. Remove pods using CRI to ensure new pods are created after kubelet restart.
7. Restart kubelet.
8. Wait for the sample device plugin pod to be running. In this case,
   the registration is not triggered.
9. Ensure that resource capacity/allocatable exported by the device plugin is zero.
10. The test pod should fail with `UnexpectedAdmissionError`
11. Delete the test pod.
12. Delete the sample device plugin pod.
13. Remove `/var/lib/kubelet/device-plugins/sample/` and its content, the directory
    created to control registration

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal commented:
/cc @fromanirh @klueska

yanirq commented on test/e2e_node/device_manager_test.go:
```diff
@@ -266,6 +272,181 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+
+			var devicePluginPodAfterRestart, tmpPod *v1.Pod
+
+			devicePluginPodAfterRestart, err = e2epod.NewPodClient(f).Get(ctx, devicePluginPod.Name, metav1.GetOptions{})
```
lint error: ineffectual assignment to err (ineffassign)

yanirq commented:
/test pull-kubernetes-unit

yanirq commented:
/test pull-kubernetes-node-e2e-containerd

Tal-or commented on test/e2e_node/device_manager_test.go:
```diff
@@ -266,6 +272,131 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
 
 	})
+
+	ginkgo.Context("With sample device plugin", func(ctx context.Context) {
```
You should not pass context as an argument to the anonymous function.
From the `node-e2e-containerd` CI lane failure:
```
W1227 11:58:05.859]   Invalid Function
W1227 11:58:05.859]   test/e2e_node/device_manager_test.go:276
W1227 11:58:05.859]     [Container] node must be passed func() - i.e. functions that take nothing and
W1227 11:58:05.860]     return nothing.  You passed func(context.Context) instead.
W1227 11:58:05.860] 
W1227 11:58:05.860]     Learn more at: http://onsi.github.io/ginkgo/#node-decorators-overview
```

Tal-or commented on test/e2e_node/device_manager_test.go:
```diff
@@ -266,6 +272,131 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+		var deviceCount int = 2
+		var devicePluginPod *v1.Pod
+		var testPods []*v1.Pod
+
```
add `ctx := context.Background()` in order to use context across the test

bart0sh commented:
/release-note-none

klueska commented:
The description says this is still WIP. Please let me know when it it ready for me to do a final review.

swatisehgal commented:
> The description says this is still WIP. Please let me know when it it ready for me to do a final review.

Thanks @klueska for pointing that out. I have updated the description as tests have been added and the  PR is ready for review.

SergeyKanzhelev commented:
/assign @fromanirh

Can you do initial review?

ffromani commented:
> /assign @fromanirh
> 
> Can you do initial review?

on it (Renamed fromanirh -> ffromani)

ffromani commented:
/cc ffromani

ffromani commented:
/cc @ffromani

ffromani commented on test/images/sample-device-plugin/sampledeviceplugin.go:
```diff
@@ -84,14 +92,59 @@ func main() {
+		}
+
+		defer conn.Close()
+		io.Copy(conn, conn)
```
why is this needed?

swatisehgal replied:
I recall playing around with handling connections in separate goroutines but later made changes and this might be a reminiscence of that. Will remove it.

ffromani commented on test/e2e_node/image_list.go:
```diff
@@ -47,7 +47,8 @@ const (
 	// SampleDevicePluginDSYAML is the path of the daemonset template of the sample device plugin. // TODO: Parametrize it by making it a feature in TestFramework.
-	SampleDevicePluginDSYAML = "test/e2e/testing-manifests/sample-device-plugin.yaml"
+	SampleDevicePluginDSYAML  = "test/e2e/testing-manifests/sample-device-plugin/sample-device-plugin-auto-registration.yaml"
+	SampleDevicePluginDS2YAML = "test/e2e/testing-manifests/sample-device-plugin/sample-device-plugin-control-registration.yaml"
```
`DS2` is a bit cryptic :) Perhaps `SampleDevicePluginControlRegistrationDSYAML`?
In general let's pick a name which is a bit more explicit about the purpose, at cost of being longer.
I think it's ok to keep the old name to cover the simpler existing cases on which the auto-registering DP is fine: tests that will need this amount of control are gonna be fewer.

swatisehgal replied:
Sure, will update.

ffromani commented on test/e2e/testing-manifests/sample-device-plugin/sample-device-plugin-control-registration.yaml:
```diff
@@ -0,0 +1,52 @@
+        hostPath:
+          path: /dev
+      containers:
+      - image: registry.k8s.io/e2e-test-images/sample-device-plugin:1.3
```
I think we will need to bump the device plugin version, which can complicate the matters, but let's check.

swatisehgal replied:
I am not very familiar with the sample-device-plugin image build process, will look into that. We would have to make sure that the image gets built with right tag is provided here.

ffromani replied:
ok, the trick here is that tests will consume the existing image already pushed in the registry. It's possible we will need to split the related commits to its own PR, then once merged, push a new sample-device-plugin image and finally consume it in the e2e tests here.
I'm not sure we can just bundle the changes and consume them in the same PR. Let's check.

swatisehgal replied:
Yes, that could be the case here. I will look into how this was done previously.

swatisehgal replied:
I will bump the sample device plugin version, point to that in the manifests and let's see how the CI reacts. That should give us an indication if the changes related to sample device plugin should be moved to a separate PR.

ffromani commented on test/e2e_node/device_manager_test.go:
```diff
@@ -266,6 +272,131 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+			}, 5*time.Minute, framework.Poll).Should(gomega.BeTrue())
+		})
+
+		ginkgo.It("should recover device plugin pod first, then pod consuming devices", func() {
```
disclosure: I authored the original commit.
The problem here is handling races. I didn't find a good way to handle them, thus give a reliable signal with this test. Let's see how subsequent commits address this problem.

swatisehgal replied:
I  summarize the steps involved in the e2e test implementation in the commit message. Let me know if I can help in better understanding here.

ffromani reviewed (commented):
initial review. I need to spend more time digesting the e2e test

swatisehgal commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -544,16 +544,25 @@ func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, requi
+	if needed == 0 {
+		// No change, no work.
+		return nil, nil
+	}
```
I am open to pursuing alternative approaches. I wanted us to have a starting point to convey the idea with the least invasive approach to tackle the issue. Will have a bit more about this in the coming days.

swatisehgal commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -544,16 +544,25 @@ func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, requi
+
 	// Check if resource registered with devicemanager
-	if _, ok := m.healthyDevices[resource]; !ok {
+	if !hasRegistered {
```
Sure, we can do that and these are feasible options option but wanted to propose the least invasive change and go from there.

I believe, above and near `isDevicePluginResource` shouldn't be any problem but if we perform this check _in_  the `isDevicePluginResource`, we would be modifying the semantics and there are a few considerations:

1. If we return false for a device plugin resource that had previously registered and are not healthy, we are silently changing the semantics of `isDevicePluginResource`. The problem with this approach could be that  we are changing the semantics of this function and with the current implementation the pod would silently skip resource allocation rather than an explicit admission error (as it also includes cases corresponding to extended resources). 

1. If we modify `isDevicePluginResource`  to return an err and add the check there we should be okay. This way the case would manifest itself as an admission error but  we would have to perform this check at multiple places but we are expanding the scope from just allocation to hint generation logic as well which is not a problem but it is something worth noting.

swatisehgal commented on test/images/sample-device-plugin/sampledeviceplugin.go:
```diff
@@ -84,14 +92,59 @@ func main() {
 		return
 	}
+
+	autoRegisterString := os.Getenv("AUTO_REGISTER")
+	if autoRegisterString == "" {
+		autoRegister = true
+	} else {
+
+		autoRegister, err = strconv.ParseBool(autoRegisterString)
+		if err != nil {
+			klog.Errorf("invalid boolean value specified")
+			panic(err)
+		}
+
+	}
```
Absolutely, Thanks!

ffromani commented on test/e2e_node/device_manager_test.go:
```diff
@@ -302,6 +302,17 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
 			devicePluginPod = e2epod.NewPodClient(f).CreateSync(ctx, dp)
 
+			go func() {
+				// Write to the Unix socket to trigger registration in a separate thread
```
let's spend words describing "why" (if not obvious) rather than "how"

ffromani commented on test/e2e_node/device_manager_test.go:
```diff
@@ -319,45 +330,20 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
-		})
-
-		ginkgo.It("should recover device plugin pod first, then pod consuming devices", func() {
+		ginkgo.It("should recover device plugin pod first, then pod consuming devices", func(ctx context.Context) {
```
is the `It` description still faithful to the intent of the code?

swatisehgal replied:
Thanks for identifying this, will fix.

ffromani commented on test/e2e_node/device_manager_test.go:
```diff
@@ -384,15 +370,78 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+			devicePluginPodAfterRestart, err = e2epod.NewPodClient(f).Get(ctx, devicePluginPod.Name, metav1.GetOptions{})
+			framework.ExpectNoError(err)
+
+			err = e2epod.WaitForPodCondition(ctx, f.ClientSet, devicePluginPodAfterRestart.Namespace, testPod.Name, "Ready", 120*time.Second, testutils.PodRunningReady)
```
is this correct? we are mixing Namespace and Name and is not obvious why. Perhaps you meant to use `devicePluginPodAfterRestart.Name` here?

swatisehgal replied:
That's right! Thanks.

ffromani commented on test/e2e_node/device_manager_test.go:
```diff
@@ -384,15 +370,78 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+				node, ready := getLocalTestNode(ctx, f)
+				return ready &&
+					numberOfDevicesCapacity(node, resourceName) == devsLen &&
+					numberOfDevicesAllocatable(node, resourceName) == devsLen
```
this looks correct, but I'd use explicit 0 to be more obvious

swatisehgal replied:
Sure, will do that.

ffromani commented on test/e2e_node/device_manager_test.go:
```diff
@@ -384,15 +370,78 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
-			}, time.Minute, time.Second).Should(gomega.BeTrue())
+			}, 2*time.Minute, time.Second).Should(gomega.BeTrue())
 
 			// TODO: here we need to tolerate pods waiting to be recreated
```
not sure this TODO is still relevant

swatisehgal replied:
Yes, it needs to be removed.

ffromani commented on test/e2e_node/device_manager_test.go:
```diff
@@ -384,15 +370,78 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+					numberOfDevicesAllocatable(node, resourceName) == devsLen
+			}, 30*time.Second, framework.Poll).Should(gomega.BeTrue())
+
+			ginkgo.By("Checking that pod requesting devices failed to start because of admission error")
```
I'd add a comment to emphasize that the registration won't happen again. We did it (racingly?) in the `beforeeach` block, which runs once the registration of the restarted device plugin won't happen unless we explicitely issue the command - and we don't want to!

swatisehgal replied:
Sure! Will do.

ffromani commented on test/e2e_node/device_manager_test.go:
```diff
@@ -384,15 +370,78 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+					return false
+				}
+
+				if !strings.Contains(tmpPod.Status.Message, "Allocate failed due to can't allocate unhealthy devices") {
```
this is a bit ugly and fragile, but I don't have better suggestions at this point :\

swatisehgal replied:
Yeah, unfortunately. I couldn't come up with a better solution for this either.

ffromani commented on test/e2e_node/device_manager_test.go:
```diff
@@ -384,15 +370,78 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+			}, 30*time.Second, framework.Poll).Should(gomega.BeTrue())
+
+			ginkgo.By("Checking that pod requesting devices failed to start because of admission error")
+			gomega.Eventually(ctx, func() bool {
```
we should also probably spend few lines explaining what is the expected scenario is and how it unfolds. For example:
we now have exactly two pods running, device plugin pod and test pod. We can't control the recovery order.
if the test pod is recovered before the device plugin pod, great: that's exactly the scenario we want to exercise.
If the device plugin pod is recovered before the test pod, we still can meaningfully reproduce the scenario by NOT sending the registration command.

swatisehgal replied:
Sure, I will explain this in more detail.

ffromani commented on test/e2e_node/device_manager_test.go:
```diff
@@ -483,41 +532,10 @@ func stringifyContainerDevices(devs []*kubeletpodresourcesv1.ContainerDevices) s
 	podName := "device-manager-test-" + string(uuid.NewUUID())
 	rl := v1.ResourceList{
-		v1.ResourceName(resourceName): *resource.NewQuantity(1, resource.DecimalSI),
+		v1.ResourceName(resourceName): *resource.NewQuantity(2, resource.DecimalSI),
```
could you please elaborate why we need 2 here?

swatisehgal replied:
We don't HAVE to provide it 2 resources but since the the sample device plugin exposes 2 devices and we have a single testpod, I give it all the resources.  

We had requested a single device in the previous commit ([5919f](https://github.com/kubernetes/kubernetes/pull/114640/commits/5919faacad20bff699c95b632875833f99773dd8)) where 2 test pods were created requesting a device each. But for this e2e a single test pod is enough.

ffromani reviewed (commented):
completed the review. The pending question now is if this fix can have unintended consequences, which I think boil down to have previosuly silently broken recovery now explicitely broken. I would call this an improvement, but still can cause some surprises.

swatisehgal commented:
> completed the review. The pending question now is if this fix can have unintended consequences, which I think boil down to have previosuly silently broken recovery now explicitely broken. I would call this an improvement, but still can cause some surprises.

This is a valid point. Perhaps we should update the release note with this information to avoid surprises.

ffromani reviewed (commented):
/lgtm
Few inline non-blocking comments. I can't think of any improvement to better tackle this issue.
The only missing item is: what's the best way to make changes to the sample device plugin? do we need to split them to a separate PR, have it merged, update the image in the registry and only then consume the changes in the e2e test? I dunno myself

swatisehgal commented on test/e2e_node/device_manager_test.go:
```diff
@@ -328,7 +328,7 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
 
 			ginkgo.By("Scheduling a sample device plugin pod")
-			data, err := e2etestfiles.Read(SampleDevicePluginDS2YAML)
+			data, err := e2etestfiles.Read(SampleDevicePluginControlRegistrationDSYAML)
```
Ack, will do!

bart0sh commented:
/assign @klueska @derekwaynecarr @mrunalp @dchen1107

swatisehgal commented on test/images/sample-device-plugin/sampledeviceplugin.go:
```diff
@@ -84,14 +91,54 @@ func main() {
+		}
+		defer l.Close()
+
+		sigc := make(chan os.Signal, 1)
+		signal.Notify(sigc, os.Interrupt, syscall.SIGTERM)
+		go func(ln net.Listener, c chan os.Signal) {
+			sig := <-c
+			klog.Infof("Received termination signal %s: shutting down.", sig)
+			ln.Close()
+			os.Exit(0)
+		}(l, sigc)
```
Added it primarily to comply with best practices.

swatisehgal commented:
> The only missing item is: what's the best way to make changes to the sample device plugin? do we need to split them to a separate PR, have it merged, update the image in the registry and only then consume the changes in the e2e test? I dunno myself

I don't understand the process about pushing images in great detail but after a bit of digging came across an [Image Promoter KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-release/1734-k8s-image-promoter) that describes the image promotion [process](https://github.com/kubernetes/enhancements/tree/master/keps/sig-release/1734-k8s-image-promoter#promotion-process) which says that that maintainers are supposed to submit a PR to add a new image. I am not sure at what stage that happens and would need some guidance here.

Splitting this into two PRs is not a problem given that we are preserving the default behavior (auto-registration). My concern is that sample device plugin change made here is motivated by being able to reproduce this bug and without the rationale captured and solved in this bug, the changes independently might be hard to digest and justify?

Since splitting this into two PRs is straightforward, I will reorganized the commits and move the sample device plugin changes to a separate PR. We can proceed further based on approach suggested by other reviewers/approvers.

ffromani commented:
> > The only missing item is: what's the best way to make changes to the sample device plugin? do we need to split them to a separate PR, have it merged, update the image in the registry and only then consume the changes in the e2e test? I dunno myself
> 
> I don't understand the process about pushing images in great detail but after a bit of digging came across an [Image Promoter KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-release/1734-k8s-image-promoter) that describes the image promotion [process](https://github.com/kubernetes/enhancements/tree/master/keps/sig-release/1734-k8s-image-promoter#promotion-process) which says that that maintainers are supposed to submit a PR to add a new image. I am not sure at what stage that happens and would need some guidance here.
> 
> Splitting this into two PRs is not a problem given that we are preserving the default behavior (auto-registration). My concern is that sample device plugin change made here is motivated by being able to reproduce this bug and without the rationale captured and solved in this bug, the changes independently might be hard to digest and justify?
> 
> Since splitting this into two PRs is straightforward, I will reorganized the commits and move the sample device plugin changes to a separate PR. We can proceed further based on approach suggested by other reviewers/approvers.

SGTM, thanks for look into this.

swatisehgal commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

bart0sh commented on test/e2e_node/device_manager_test.go:
```diff
@@ -266,6 +272,231 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+		spec with AUTO_REGISTER=false to first manually register the plugin to allow
+		an application pod (requesting devices) to successfully run on the node
+		followed by kubelet restart where device plugin doesn't register and the
+		application pod fails with admission error.
```
Can we achieve the same result by not starting device plugin after the kubelet restart?

swatisehgal replied:
The device plugin and the application pod are recovered automatically after kubelet restart. We cannot influence the device plugin pod start-up as the order in which the pods are recovered is not in our control.

bart0sh replied:
Now it makes sense to me. Thanks for the explanations.

swatisehgal replied:
No problem!

klueska commented on pkg/kubelet/cm/devicemanager/manager.go:
```diff
@@ -544,16 +544,25 @@ func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, requi
-		return nil, nil
-	}
+
 	klog.V(3).InfoS("Need devices to allocate for pod", "deviceNumber", needed, "resourceName", resource, "podUID", string(podUID), "containerName", contName)
+	healthyDevices, hasRegistered := m.healthyDevices[resource]
+
 	// Check if resource registered with devicemanager
-	if _, ok := m.healthyDevices[resource]; !ok {
+	if !hasRegistered {
 		return nil, fmt.Errorf("can't allocate unregistered device %s", resource)
 	}
 
+	// Check if registered resource has healthy devices
+	if healthyDevices.Len() == 0 {
+		return nil, fmt.Errorf("can't allocate unhealthy devices %s", resource)
+	}
+
+	if needed == 0 {
+		// No change, no work.
+		return nil, nil
+	}
+
```
As far as I can tell, this is the only "core" change (and it seems reasonable / logical).
The rest are changes to testing, is that correct?

klueska replied:
Actually, I have one question here -- your PR description mentioned that you wanted to check that all already-allocated devices were in fact healthy (and error out otherwise). I don't see this logic here. What you have in this PR is strictly better than what we had before, but I'm curious if this check was left out on purpose or overlooked.

swatisehgal replied:
> As far as I can tell, this is the only "core" change (and it seems reasonable / logical).
The rest are changes to testing, is that correct?

Yes, this is accurate. Please note that https://github.com/kubernetes/kubernetes/pull/115107 would need to merge first as we need that for e2e tests implementation here. Also, due to some suggestions on that PR I have proposed a file based synchronization approach (as [3rd commit](https://github.com/kubernetes/kubernetes/pull/115107/commits/2aabf15c370ca727067f550a7ff225ec6d85af57) on that PR) diverging slightly sample device plugin changes in these two PRs temporarily. Once we reach consensus on that, I will rebase.

> Actually, I have one question here -- your PR description mentioned that you wanted to check that all already-allocated devices were in fact healthy (and error out otherwise). I don't see this logic here. What you have in this PR is strictly better than what we had before, but I'm curious if this check was left out on purpose or overlooked.

The change here: 
https://github.com/kubernetes/kubernetes/pull/114640/commits/6d80d27b8204b32b82da4e6ad3c10855327d36b0#diff-623f1164bd7d648ee0421742187c76b5071b44315aee695f29edc0203b6c762dR556-R559 is implementing that logic.
Previously, while iterating through `m.healthyDevices` we **only** checked if devices had registered which meant that it was okay to proceed forward with allocation if we had devices that previously registered even if they are unhealthy (and therefore unsuitable to be allocated after reboot either because the device plugin pod didn't started or didn't register itself yet to kubelet) .
Now, we obtain the list of `healthyDevices`, check its length to validate the presence of healthy devices and rule out `healthyDevices` with zero length as that is the case that corresponds to registered **unhealthy** devices. We only proceed to allocation if both these conditions are met ensuring that these devices are actually healthy and don't just appear to be because at some stage they had registered to kubelet.

klueska replied:
Maybe I misinterpreted what you meant then.

I thought you were saying you planned to validate that all of the already-allocated devices (i.e. those present in the `devices` variable at this point in the code) are healthy. The check you have in place just ensures that "at least one" device is healthy, but not that the full set of already-allocated devices are all healthy.

swatisehgal replied:
I see what you mean, we can add a check to make sure that the `devices` retrieved from the checkpoint file are healthy as well.

I had been working under the assumption that nothing changes in the environment and once the device plugin has registered we would have enough devices to fulfill the pod resource requirement. I now see you have a valid point, we could have a case where a device plugin might have  re-registered but there aren't enough healthy devices e.g. a node had two NICs and devices are allocated from two NIC to a pod but one of them was removed before node reboot.

We can improve the current implementation by checking that there are at least `devices.Len()` number of devices instead of the current naive check of "at least one" healthy device. Also, we would have to check if `devices` is a subset of `healthyDevices`.

The only case that I want like to explicitly point out is where devices previously allocated (and thus retrieved from the checkpoint) are not healthy anymore but there are enough devices that can be allocated to the pod. In that case as well we would be returning an admission error. I think that is fair because without that we would proceed as normal but the pod would fail when it will try to access unhealthy devices "allocated" to it.

Let me know if that makes sense.

swatisehgal replied:
Thought about this a bit more...
After node reboot, for all the registered devices  `m.healthyDevices` field is reset to zero because of which device capacity is 0. This is intentional as we want to prevent additional pods requesting devices to be scheduled on the node.  

Since `devices` is not empty, we check if additional resources need to be allocated`needed = needed - devices.Len()`

it is sufficient to check if healthyDevices.len is zero as that is an indication of the device plugin not being registered.

Since `healthyDevices` is empty we can't cross check if the devices retrieved are healthy or not.

klueska replied:
I was talking about the state of things after ensuring that healthyDevices.Len() > 0 (and consequently before checking if needed == 0).

The set of already-allocated-devices at this point are just assumed to all be healthy if we don’t explicitly check for this. 

I’m just wondering if we want to cross check them against the set of devices in healthyDevices and if they are not all there, we error out.

swatisehgal replied:
I see, I will handle that case as well. Thanks.

swatisehgal committed:
node: device-mgr: explicitly check if pre-allocated devices are healthy

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal commented:
/retest

klueska commented:
/approve
/hold

The core change looks good to me. I haven't looked closely at the updates to the tests, but if other sig-node reviewers are happy with the changes there, feel free to add an LGTM and remove the hold.

swatisehgal commented:
> /approve /hold
> 
> The core change looks good to me. I haven't looked closely at the updates to the tests, but if other sig-node reviewers are happy with the changes there, feel free to add an LGTM and remove the hold.

Thanks @klueska!

/assign @pohly @aojea
for review/approval on test related changes

pohly commented on test/e2e_node/device_manager_test.go:
```diff
@@ -266,6 +272,231 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+			gomega.Eventually(func() bool {
+				node, ready := getLocalTestNode(ctx, f)
+				return ready && numberOfSampleResources(node) > 0
+			}, 5*time.Minute, framework.Poll).Should(gomega.BeTrue())
```
`BeTrue` is not going to be informative. You can use https://pkg.go.dev/github.com/onsi/gomega/gcustom to write a better one for this particular check.

swatisehgal replied:
Can you elaborate on what you would like to be done here. You want me to write up a custom matcher here?

We come to the conclusion that a node is ready if:

1. node status is ready
2. network is setup ( which is done by checking `NetworkUnavailable` condition is not true) 
3. node is not unschedulable
4. device plugin advertises resources on the node

Point 1-3 are exactly what [getLocatTestNode](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/util.go#L263-L270) does for us. I know it is not ideal that this is returned as a boolean and it would probably be better to have a custom matcher that allows us to `.Should(BeReady)` but that would mean we are reimplmenting a lot of code and I am not sure if that is the best approach.

pohly replied:
I see that [getLocatTestNode](https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/util.go#L263-L270) is basically a combination of "get node",  `e2enode.IsNodeReady` and `e2enode.IsNodeSchedulable`.

A good solution would be to:
- get the node in the `Eventually` callback
- rewrite `IsNodeReady` and `IsNodeSchedulable` as gomega matchers
- use `gomega.SatisfyAll` with those

But I understand that this is probably more work than you can afford or want to invest.

Some intermediate solution would be to get the node, then use a matcher based on gcustom which invokes `getLocatTestNode` with a failure message that explains the boolean.

pohly replied:
You can look at e2epod for inspiration. It was recently rewritten to use such a pattern.

pohly reviewed (changes requested):
I pointed out some areas where the tests could be improved. The same comments apply in multiple places, I stopped repeating them at some point.

swatisehgal committed:
node: device-mgr: e2e: address e2e test review comments

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal commented on test/e2e_node/device_manager_test.go:
```diff
@@ -266,6 +272,231 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+		// this test wants to reproduce what happened in https://github.com/kubernetes/kubernetes/issues/109595
+		ginkgo.BeforeEach(func(ctx context.Context) {
+			ginkgo.By("Wait for node to be ready")
+			gomega.Eventually(func() bool {
```
Ack, Thanks!

swatisehgal commented on test/e2e_node/device_manager_test.go:
```diff
@@ -266,6 +272,231 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+			devicePluginPod = e2epod.NewPodClient(f).CreateSync(ctx, dp)
+
+			go func() {
+				// Since autoregistration is disabled for the device plugin (as AUTO_REGISTER=false),
```
Thanks, will address.

swatisehgal commented on test/e2e_node/device_manager_test.go:
```diff
@@ -266,6 +272,231 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+
+				triggerPath := devicePluginDir + "/registered"
+				conn, err := net.Dial("unix", triggerPath)
+				framework.ExpectNoError(err)
```
Yes, that is what I was hoping for.

swatisehgal commented on test/e2e_node/device_manager_test.go:
```diff
@@ -266,6 +272,231 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+				nodes, err := e2enode.TotalReady(ctx, f.ClientSet)
+				framework.ExpectNoError(err)
+				return nodes == 1
+			}, time.Minute, time.Second).Should(gomega.BeTrue())
```
Thanks, I will address this.

pohly commented on test/e2e_node/device_manager_test.go:
```diff
@@ -351,3 +579,98 @@ func stringifyContainerDevices(devs []*kubeletpodresourcesv1.ContainerDevices) s
+}
+
+// HaveAllocatableDevices verifies that a node has allocatable devices.
+func HaveAllocatableDevices(ctx context.Context, f *framework.Framework) types.GomegaMatcher {
```
The parameters are unused?

swatisehgal replied:
Right, will update.

pohly commented on test/e2e_node/device_manager_test.go:
```diff
@@ -351,3 +579,98 @@ func stringifyContainerDevices(devs []*kubeletpodresourcesv1.ContainerDevices) s
+}
+
+// BeReady verifies that a node is ready and devices have registered.
+func BeReady(ctx context.Context, f *framework.Framework) types.GomegaMatcher {
```
The parameters are unused?

pohly commented on test/e2e_node/device_manager_test.go:
```diff
@@ -266,6 +273,227 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+			gomega.Eventually(ctx, func(ctx context.Context) bool {
+				node, ready := getLocalTestNode(ctx, f)
+				return ready && numberOfSampleResources(node) <= 0
+			}, 5*time.Minute, framework.Poll).Should(gomega.BeTrue())
```
Here's another candidate for replacing `gomega.BeTrue` with a matcher :sweat_smile:

swatisehgal replied:
I missed this one, will update now.

pohly commented on test/e2e_node/device_manager_test.go:
```diff
@@ -266,6 +273,227 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+
+				return true
+			}, time.Minute, 5*time.Second).Should(
+				gomega.Equal(true),
```
Here's another candidate for replacing gomega.BeTrue with a matcher sweat_smile

swatisehgal commented:
Force pushed as I rebased the PR.

swatisehgal committed:
node: device-mgr: e2e: adapt to sample device plugin refactoring

These updates are to adapt to the sample device plugin
refactoring done here: https://github.com/kubernetes/kubernetes/pull/115926/commits/92e00203e027bf88ddc695fc17e80c913a6b178a.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

pohly commented on test/e2e_node/device_manager_test.go:
```diff
@@ -266,6 +273,211 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
+			ginkgo.By("Waiting for devices to become available on the local node")
+
+			gomega.Eventually(ctx, isNodeReadyWithSampleResources).
+				WithArguments(ctx, f).
```
Don't pass `ctx` as argument. `gomega.Eventually` will handle that automatically.

I find it easier to write:
```
gomega.Eventually(ctx, isNodeReadyWithSampleResources, f).
```
without the `WithArguments`

pohly replied:
This also applies below.

swatisehgal replied:
Thanks, I am not passing `ctx` as argument in the `WithArguments` anymore. 

I tried playing around with the approach of not using `WithArguments` as per your suggestion but the test fails without it.

pohly replied:
Fails how? I'm curious because the approach with directly passing the arguments is used everywhere else, `WithArguments` is not used anywhere.

swatisehgal replied:
Yep, I thought so too when I referred to other areas in the code base that used eventually the way you suggested. Here is what I get:

```
  Feb 28 19:06:07.615: INFO: 
  Logging kubelet events for node swsehgal-node
  Feb 28 19:06:07.618: INFO: 
  Logging pods the kubelet thinks is on node swsehgal-node
  W0228 19:06:07.623862  111282 metrics_grabber.go:111] Can't find any pods in namespace kube-system to grab metrics from
  Feb 28 19:06:07.657: INFO: 
  Latency metrics for node swsehgal-node
  STEP: Destroying namespace "devicemanager-test-3328" for this suite. @ 02/28/23 19:06:07.658
  << Timeline

  [FAILED] &framework.Framework{BaseName:"devicemanager-test", UniqueName:"devicemanager-test-3328", clientConfig:(*rest.Config)(0xc001ab0240), ClientSet:(*kubernetes.Clientset)(0xc0001a0b60), KubemarkExternalClusterClientSet:kubernetes.Interface(nil), DynamicClient:(*dynamic.DynamicClient)(0xc00061d570), ScalesGetter:(*scale.scaleClient)(0xc001b09a80), SkipNamespaceCreation:false, Namespace:(*v1.Namespace)(0xc000f68420), namespacesToDelete:[]*v1.Namespace{(*v1.Namespace)(0xc000f68420)}, NamespaceDeletionTimeout:0, NamespacePodSecurityEnforceLevel:"privileged", flakeReport:(*framework.FlakeReport)(0xc001b09dc0), Options:framework.Options{ClientQPS:20, ClientBurst:50, GroupVersion:(*schema.GroupVersion)(nil)}, TestSummaries:[]framework.TestDataSummary(nil), Timeouts:(*framework.TimeoutContext)(0xc000ecb180), DumpAllNamespaceInfo:(framework.DumpAllNamespaceInfoAction)(0x4679320)} is not a valid interval. Must be a time.Duration, a parsable duration string, or a number.
  In [BeforeEach] at: test/e2e_node/device_manager_test.go:386 @ 02/28/23 19:06:03.575
------------------------------
SSSSSSSSSSSSS

Summarizing 1 Failure:
  [FAIL] [sig-node] Device Manager  [Serial] [Feature:DeviceManagerBug][NodeFeature:DeviceManager] With sample device plugin [BeforeEach] should deploy pod consuming devices first but fail with admission error after kubelet restart in case device plugin hasn't re-registered
  test/e2e_node/device_manager_test.go:386

Ran 1 of 389 Specs in 16.926 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 388 Skipped
--- FAIL: TestE2eNode (16.96s)
```

swatisehgal replied:
The only change I made was:
```
diff --git a/test/e2e_node/device_manager_test.go b/test/e2e_node/device_manager_test.go
index 68eeb85b2d2..6fc60fc5ed0 100644
--- a/test/e2e_node/device_manager_test.go
+++ b/test/e2e_node/device_manager_test.go
@@ -383,8 +383,7 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
 
                        ginkgo.By("Waiting for devices to become available on the local node")
 
-                       gomega.Eventually(ctx, isNodeReadyWithSampleResources).
-                               WithArguments(f).
+                       gomega.Eventually(ctx, isNodeReadyWithSampleResources, f).
                                WithTimeout(5 * time.Minute).
                                Should(BeReady())
```

pohly replied:
I was wrong about  `Eventually` taking a function + parameters. The additional arguments after the function are timeout and poll interval. Your approach with `WithArguments` is correct.

swatisehgal replied:
Thanks for double checking this.

pohly reviewed (changes requested):
One small nit, otherwise very nice usage of `gomega.Eventually`.

But someone else will have to LGTM, there are too many parts that I am not familiar with.

swatisehgal commented:
> But someone else will have to LGTM, there are too many parts that I am not familiar with

@klueska reviewed this PR and [approved](https://github.com/kubernetes/kubernetes/pull/114640#issuecomment-1415840405) the core changes.

@fromani reviewed this PR gave his [LGTM](https://github.com/kubernetes/kubernetes/pull/114640#pullrequestreview-1247388149). It got removed as I was addressing some review comments and had to rebase.

With regard to tackling the consumption of sample device plugin changes in e2e tests here, the sample device plugin changes were split into a separate [PR](https://github.com/kubernetes/kubernetes/pull/115107) which would need to merge first. @bart0sh has been reviewing that PR and gave his [LGTM](https://github.com/kubernetes/kubernetes/pull/115107#pullrequestreview-1286820735) and indicated that he is [okay](https://github.com/kubernetes/kubernetes/pull/115107#issuecomment-1438932149) with the approach taken.

If you are happy enough with these these two PRs from test perspective and can approve, we should be able to move forward.

ffromani commented:
> @fromani reviewed this PR gave his [LGTM](https://github.com/kubernetes/kubernetes/pull/114640#pullrequestreview-1247388149). It got removed as I was addressing some review comments and had to rebase.

And I'll review again before end of day (EOD)

swatisehgal commented:
> > @fromani reviewed this PR gave his [LGTM](https://github.com/kubernetes/kubernetes/pull/114640#pullrequestreview-1247388149). It got removed as I was addressing some review comments and had to rebase.
> 
> And I'll review again before end of day (EOD)

Thank you!

ffromani reviewed (commented):
first two commits reviewed on https://github.com/kubernetes/kubernetes/pull/115107 - conditional LGTM here, but they need to be merged separately anyway
next three commits already approved, and LGTM anyway
the rest of the commits are about e2e changes which LGTM - nice use of custom gomega matchers!

ffromani commented:
/lgtm

for the test changes.
*please* don't remove the hold just yet: for this PR to merge we need to merge https://github.com/kubernetes/kubernetes/pull/115107 and promote the sample device plugin image. I see the approval was removed somehow (!?) so it should be OK.

ffromani commented:
> /lgtm
> 
> for the test changes. _please_ don't remove the hold just yet: for this PR to merge we need to merge #115107 and promote the sample device plugin image. I see the approval was removed somehow (!?) so it should be OK.

proof:
![image](https://user-images.githubusercontent.com/25349329/221959119-8f0554f0-a9bf-474e-9747-d81e9b98893c.png)

swatisehgal commented:
> /lgtm
> 
> for the test changes. _please_ don't remove the hold just yet: for this PR to merge we need to merge #115107 and promote the sample device plugin image.

Thanks! Correct, the PR with sample device changes has to merge first.

> I see the approval was removed somehow (!?) so it should be OK.

I'm not 100% sure but maybe the label was never added because we need approval from both kubelet owners and test owners. See the image below:
![image](https://user-images.githubusercontent.com/18619641/222108174-c8b8509e-b72d-4c7b-ac35-c13434a2ef7e.png)

ffromani commented on test/e2e_node/device_manager_test.go:
```diff
@@ -287,7 +327,7 @@ var _ = SIGDescribe("Device Manager  [Serial] [Feature:DeviceManager][NodeFeatur
 
 			ginkgo.By("Scheduling a sample device plugin pod")
-			data, err := e2etestfiles.Read(SampleDevicePluginDS2YAML)
+			data, err := e2etestfiles.Read(SampleDevicePluginControlRegistrationDSYAML)
```
this manifest correctly uses the 1.5 plugin version, which is not promoted yet. This is just fine for now. Let's promote the plugin and we can move forward.

ffromani replied:
Thanks for getting the image promoted (https://github.com/kubernetes/k8s.io/pull/4862). I'll run a final check very shortly and review (and likely LGTM) again. I see no issues at all, just the last bit of due diligence.

ffromani reviewed (commented):
Thanks for the rebase! LGTM. Holding the formal LGTM label until the new sample-device-plugin image is promoted. Ping me anytime once this happens so I can tag this PR.

ffromani commented:
/test pull-kubernetes-unit

apiserver failures can't be cause by this PR

ffromani commented:
@swatisehgal ok, the last check found a couple issues. Good news is I think there are both easy to fix.
1. I know you made changes to make to update the e2e test work with the latest iteration (which got merged) of the sample device plugin, moving all the file handling in the e2e test (previously it was mixed between the plugin and the e2e test). Could you please check this PR is up to date and includes all your recent work?
2. we took most care to use the newer sampledevice plugin ONLY in this new test to minimize the chance of regressions, deferring cleanups to later PR. I think this is the right call. Unfortunately we missed an update, I think the following should be enough to fix:
```
diff --git a/test/e2e_node/image_list.go b/test/e2e_node/image_list.go
index 611590642bc..e6e7da29766 100644
--- a/test/e2e_node/image_list.go
+++ b/test/e2e_node/image_list.go
@@ -87,11 +87,16 @@ func updateImageAllowList(ctx context.Context) {
        } else {
                e2epod.ImagePrePullList.Insert(gpuDevicePluginImage)
        }
-       if samplePluginImage, err := getSampleDevicePluginImage(); err != nil {
+       if samplePluginImage, err := getContainerImageFromE2ETestDaemonset(SampleDevicePluginDSYAML); err != nil {
                klog.Errorln(err)
        } else {
                e2epod.ImagePrePullList.Insert(samplePluginImage)
        }
+       if samplePluginImageCtrlReg, err := getContainerImageFromE2ETestDaemonset(SampleDevicePluginControlRegistrationDSYAML); err != nil {
+               klog.Errorln(err)
+       } else {
+               e2epod.ImagePrePullList.Insert(samplePluginImageCtrlReg)
+       }
 }
 
 func getNodeProblemDetectorImage() string {
@@ -222,19 +227,19 @@ func getGPUDevicePluginImage(ctx context.Context) (string, error) {
        return ds.Spec.Template.Spec.Containers[0].Image, nil
 }
 
-func getSampleDevicePluginImage() (string, error) {
-       data, err := e2etestfiles.Read(SampleDevicePluginDSYAML)
+func getContainerImageFromE2ETestDaemonset(dsYamlPath string) (string, error) {
+       data, err := e2etestfiles.Read(dsYamlPath)
        if err != nil {
-               return "", fmt.Errorf("failed to read the sample plugin yaml: %w", err)
+               return "", fmt.Errorf("failed to read the daemonset yaml: %w", err)
        }
 
        ds, err := e2emanifest.DaemonSetFromData(data)
        if err != nil {
-               return "", fmt.Errorf("failed to parse daemon set for sample plugin: %w", err)
+               return "", fmt.Errorf("failed to parse daemonset yaml: %w", err)
        }
 
        if len(ds.Spec.Template.Spec.Containers) < 1 {
-               return "", fmt.Errorf("failed to parse the sample plugin image: cannot extract the container from YAML")
+               return "", fmt.Errorf("failed to parse the container image: cannot extract the container from YAML")
        }
        return ds.Spec.Template.Spec.Containers[0].Image, nil
 }
```

swatisehgal commented:
Thanks @ffromani for your thorough review. Appreciate it! I have addressed your review comments.

> 1. I know you made changes to make to update the e2e test work with the latest iteration (which got merged) of the sample device plugin, moving all the file handling in the e2e test (previously it was mixed between the plugin and the e2e test). Could you please check this PR is up to date and includes all your recent work?

Thanks for identifying this. I think I must have rebased an older version of the code. Fixed it now.
> 2. we took most care to use the newer sampledevice plugin ONLY in this new test to minimize the chance of regressions, deferring cleanups to later PR. I think this is the right call. Unfortunately we missed an update, I think the following should be enough to fix:

Thanks, fixed.

ffromani commented:
/lgtm

thanks for quickly addressing my comments. Codewise looks good to me now, and the tests on my local env are now fine as well.

swatisehgal commented:
/unhold
as all the review comments have been addressed and both the prerequisite PRs (https://github.com/kubernetes/kubernetes/pull/115107 and https://github.com/kubernetes/k8s.io/pull/4862) have merged.

ffromani commented:
/test pull-kubernetes-integration

apiserver test failures can't be caused by this PR

liggitt commented:
cleared release note and reopened issue marked as fixed due to revert of this PR in https://github.com/kubernetes/kubernetes/pull/116341

swatisehgal commented:
Here is a PR with updates to ensure that the device plugin tests pass: https://github.com/kubernetes/kubernetes/pull/116376.
